# Building Logistic Regression Model for Absenteeism Project

Now that we have our preprocessed data, we can start building a model that can take in the predictors and predict whether a random person is likely to be absent at work.

In this notebook, we will be creating a Logistic Regression (Logit) model to do the job.

## Import the Necessary Libraries

In [1]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.base import BaseEstimator, TransformerMixin

## Import the Data

In [2]:
data_preprocessed = pd.read_csv('../Dataset/Absenteeism_preprocessed.csv')
data_preprocessed

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0                 30          0         2     1                          4  
1                 31          0         1     0                          0  
2                 31          0         0     0                          2  
3                 24          0         2     0                          4  
4                 30          0         2     1                          2  
..               ...        ...       ...   ...                        ...  
695               22          1         2     0                          8  
696               24          0         1     2                          3  
697               25          1         0     0                          8  
698               25          1         0     0                          2  
699               25          0         1     1                          2  

[700 rows x 15 columns]

## Create the Targets

For this project, we will be differentiating between excessive absenteeism and moderate absenteeism. For that, we will create target variable containing 1 or 0 for each depending on if the `Absenteeism Time in Hours` is more than or less than its median value respectively.

Finally, we can drop the `Absenteeism Time in Hours` column from the dataframe as it won't be needed anymore.

In [3]:
# create a target variable based on the median of `Absenteeism Time in Hours`
targets = np.where(data_preprocessed['Absenteeism Time in Hours'] > data_preprocessed['Absenteeism Time in Hours'].median(), 1, 0)
targets

array([1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,

In [4]:
# add the newly created targets to the dataframe
data_preprocessed['Excessive Absenteeism'] = targets
data_preprocessed

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  \
0                 30          0         2     1                          4   
1                 31          0         1     0                          0   
2                 31          0         0     0                          2   
3                 24          0         2     0                          4   
4                 30          0         2     1                          2   
..               ...        ...       ...   ...                        ...   
695               22          1         2     0                          8   
696               24          0         1     2                          3   
697               25          1         0     0                          8   
698               25          1         0     0                          2   
699               25          0         1     1                          2   

     Excessive Absenteeism  
0                        1  
1                        0  
2                        0  
3                        1  
4                        0  
..                     ...  
695                      1  
696                      0  
697                      1  
698                      0  
699                      0  

[700 rows x 16 columns]

In [5]:
# drop the `Absenteeism in Hours` column
data_with_targets = data_preprocessed.drop(['Absenteeism Time in Hours'], axis=1)

data_with_targets

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  Excessive Absenteeism  
0                 30          0         2     1                      1  
1                 31          0         1     0                      0  
2                 31          0         0     0                      0  
3                 24          0         2     0                      1  
4                 30          0         2     1                      0  
..               ...        ...       ...   ...                    ...  
695               22          1         2     0                      1  
696               24          0         1     2                      0  
697               25          1         0     0                      1  
698               25          1         0     0                      0  
699               25          0         1     1                      0  

[700 rows x 15 columns]

### Balancing classes

For a good classification model, it is necessary to balance the classes of the samples in the dataset. It will create more generalised model.

In [6]:
# check the proportion of the 2 classes
targets.sum() / targets.shape[0]

0.45571428571428574

As we can see, around 46% of the data belongs to class 1 (Excessive Absenteeism) while the rest 54% to class 0 (Moderate Absenteeism). Hence, the dataset is roughly balanced and does not need further interference.

## Create the Inputs

In [7]:
# inputs will be all columns except the targets
unscaled_inputs = data_with_targets.iloc[:, :-1]
unscaled_inputs

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1            7                1   
1           0         0         0         0            7                1   
2           0         0         0         1            7                2   
3           1         0         0         0            7                3   
4           0         0         0         1            7                3   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0            5                2   
696         1         0         0         0            5                2   
697         1         0         0         0            5                3   
698         0         0         0         1            5                3   
699         0         0         0         1            5                3   

     Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                       289                36   33                  239.554   
1                       118                13   50                  239.554   
2                       179                51   38                  239.554   
3                       279                 5   39                  239.554   
4                       289                36   33                  239.554   
..                      ...               ...  ...                      ...   
695                     179                22   40                  237.656   
696                     225                26   28                  237.656   
697                     330                16   28                  237.656   
698                     235                16   32                  237.656   
699                     291                31   40                  237.656   

     Body Mass Index  Education  Children  Pets  
0                 30          0         2     1  
1                 31          0         1     0  
2                 31          0         0     0  
3                 24          0         2     0  
4                 30          0         2     1  
..               ...        ...       ...   ...  
695               22          1         2     0  
696               24          0         1     2  
697               25          1         0     0  
698               25          1         0     0  
699               25          0         1     1  

[700 rows x 14 columns]

## Scale the Inputs

As we can see, the inputs vary in magnitude and units, and so can impact our model negatively. To ensure equal importance for all the features, we should scale them.

I will use the Standard Scaler such that all column values have a mean of 0 and standard deviation of 1.

I will be creating a custom scaler that does not scale the dummy variables.

In [8]:
# create the Custom Scaler class

class CustomScaler(BaseEstimator,TransformerMixin):

    # init or what information we need to declare a CustomScaler object
    # and what is calculated/declared as we do

    def __init__(self,columns,copy=True,with_mean=True,with_std=True):

        # scaler is nothing but a Standard Scaler object
        self.scaler = StandardScaler(copy=copy, with_mean=with_mean, with_std=with_std)
        # with some columns 'twist'
        self.columns = columns
        self.mean_ = None
        self.var_ = None


    # the fit method, which, again based on StandardScale

    def fit(self, X, y=None):
        self.scaler.fit(X[self.columns], y)
        self.mean_ = np.mean(X[self.columns], axis=1)
        self.var_ = np.var(X[self.columns], axis=1)
        return self

    # the transform method which does the actual scaling

    def transform(self, X, y=None, copy=None):

        # record the initial order of the columns
        init_col_order = X.columns

        # scale all features that you chose when creating the instance of the class
        X_scaled = pd.DataFrame(self.scaler.transform(X[self.columns]), columns=self.columns)

        # declare a variable containing all information that was not scaled
        X_not_scaled = X.loc[:,~X.columns.isin(self.columns)]

        # return a data frame which contains all scaled features and all 'not scaled' features
        # use the original order (that you recorded in the beginning)
        return pd.concat([X_not_scaled, X_scaled], axis=1)[init_col_order]

In [9]:
# choose the columns to scale
columns_to_scale = ['Month Value','Day of the Week', 'Transportation Expense', 'Distance to Work','Age', 'Daily Work Load Average', 'Body Mass Index', 'Children', 'Pets']

In [10]:
# declare a scaler object, specifying the columns to scale
absenteeism_scaler = CustomScaler(columns = columns_to_scale)

In [11]:
# fit the scaler to the data
absenteeism_scaler.fit(unscaled_inputs)

# transform the inputs
scaled_inputs = absenteeism_scaler.transform(unscaled_inputs)

In [12]:
scaled_inputs

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1     0.182726        -0.683704   
1           0         0         0         0     0.182726        -0.683704   
2           0         0         0         1     0.182726        -0.007725   
3           1         0         0         0     0.182726         0.668253   
4           0         0         0         1     0.182726         0.668253   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0    -0.388293        -0.007725   
696         1         0         0         0    -0.388293        -0.007725   
697         1         0         0         0    -0.388293         0.668253   
698         0         0         0         1    -0.388293         0.668253   
699         0         0         0         1    -0.388293         0.668253   

     Transportation Expense  Distance to Work       Age  \
0                  1.005844          0.412816 -0.536062   
1                 -1.574681         -1.141882  2.130803   
2                 -0.654143          1.426749  0.248310   
3                  0.854936         -1.682647  0.405184   
4                  1.005844          0.412816 -0.536062   
..                      ...               ...       ...   
695               -0.654143         -0.533522  0.562059   
696                0.040034         -0.263140 -1.320435   
697                1.624567         -0.939096 -1.320435   
698                0.190942         -0.939096 -0.692937   
699                1.036026          0.074838  0.562059   

     Daily Work Load Average  Body Mass Index  Education  Children      Pets  
0                  -0.806331         0.767431          0  0.880469  0.268487  
1                  -0.806331         1.002633          0 -0.019280 -0.589690  
2                  -0.806331         1.002633          0 -0.919030 -0.589690  
3                  -0.806331        -0.643782          0  0.880469 -0.589690  
4                  -0.806331         0.767431          0  0.880469  0.268487  
..                       ...              ...        ...       ...       ...  
695                -0.853789        -1.114186          1  0.880469 -0.589690  
696                -0.853789        -0.643782          0 -0.019280  1.126663  
697                -0.853789        -0.408580          1 -0.919030 -0.589690  
698                -0.853789        -0.408580          1 -0.919030 -0.589690  
699                -0.853789        -0.408580          0 -0.019280  0.268487  

[700 rows x 14 columns]

In [13]:
type(scaled_inputs)

pandas.core.frame.DataFrame

## Split the Dataset

We will split the dataset into training and testing in an 80:20 ratio. We will fix a random state to ensure consistency, and stratify the dataset to ensure the class balance in the splits.

In [14]:
# split the data into 80:20
# use random state = 1
# stratify the data
x_train, x_test, y_train, y_test = train_test_split(scaled_inputs, targets, test_size=0.2, stratify=targets, random_state=1)

In [15]:
# check train shape
print(x_train.shape, y_train.shape)

(560, 14) (560,)


In [16]:
# check test shape
print(x_test.shape, y_test.shape)

(140, 14) (140,)


In [17]:
# check dataset balance
print(y_train.sum() / y_train.shape[0])
print(y_test.sum() / y_test.shape[0])

0.45535714285714285
0.45714285714285713


The split was a success and the class balance has also been maintained.

## Build the Model

In [18]:
# build a logistic regression model
reg = LogisticRegression()

# train the model
reg.fit(x_train, y_train)

LogisticRegression()

In [19]:
# check the accuracy
reg.score(x_train, y_train)

0.7821428571428571

Using the model, we achieved a training accuracy of 78%.

### Manually check the accuracy

In [20]:
model_output = reg.predict(x_train)
model_output

array([0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,

In [21]:
true_val = np.sum(model_output == y_train)
acc = true_val / model_output.shape[0]
acc

0.7821428571428571

## Find the Intercept and Coefficients

In [22]:
# find intercept
reg.intercept_

array([-1.72367645])

In [23]:
# find coefficients
reg.coef_

array([[ 2.88027971,  0.92324043,  2.90326738,  0.88967415,  0.07005452,
        -0.2952236 ,  0.59927666,  0.02972708, -0.1832303 ,  0.01310325,
         0.24716318,  0.21118   ,  0.43666114, -0.22218894]])

### Create a Summary Table

A summary table will give us a better look at the coefficients and their respective features, and the intercept.

In [24]:
# create a dataframe for features and coefficients
summary_table = pd.DataFrame(data = unscaled_inputs.columns.values, columns = ['Feature Names'])

In [25]:
# add the coeff
# transpose the coeff as it has multiple columns
summary_table['Coefficients'] = np.transpose(reg.coef_)

In [26]:
summary_table

Feature Names  Coefficients
0                  Reason 1      2.880280
1                  Reason 2      0.923240
2                  Reason 3      2.903267
3                  Reason 4      0.889674
4               Month Value      0.070055
5           Day of the Week     -0.295224
6    Transportation Expense      0.599277
7          Distance to Work      0.029727
8                       Age     -0.183230
9   Daily Work Load Average      0.013103
10          Body Mass Index      0.247163
11                Education      0.211180
12                 Children      0.436661
13                     Pets     -0.222189

In [27]:
# add the intercept
# shift all rows 1 down and add at the beginning
summary_table.index = summary_table.index + 1

summary_table.loc[0] = ['Intercept', reg.intercept_[0]]

In [28]:
summary_table = summary_table.sort_index()
summary_table

Feature Names  Coefficients
0                 Intercept     -1.723676
1                  Reason 1      2.880280
2                  Reason 2      0.923240
3                  Reason 3      2.903267
4                  Reason 4      0.889674
5               Month Value      0.070055
6           Day of the Week     -0.295224
7    Transportation Expense      0.599277
8          Distance to Work      0.029727
9                       Age     -0.183230
10  Daily Work Load Average      0.013103
11          Body Mass Index      0.247163
12                Education      0.211180
13                 Children      0.436661
14                     Pets     -0.222189

## Interpret the Results

We should calculate the odds ratio to find out how much the change in each feature impact the odds of being absent.

In [29]:
# calculate the odds ratio
# append in the summary table
summary_table['Odds Ratio'] = np.exp(summary_table['Coefficients'])

In [30]:
summary_table

Feature Names  Coefficients  Odds Ratio
0                 Intercept     -1.723676    0.178409
1                  Reason 1      2.880280   17.819257
2                  Reason 2      0.923240    2.517435
3                  Reason 3      2.903267   18.233624
4                  Reason 4      0.889674    2.434336
5               Month Value      0.070055    1.072567
6           Day of the Week     -0.295224    0.744365
7    Transportation Expense      0.599277    1.820801
8          Distance to Work      0.029727    1.030173
9                       Age     -0.183230    0.832576
10  Daily Work Load Average      0.013103    1.013189
11          Body Mass Index      0.247163    1.280388
12                Education      0.211180    1.235135
13                 Children      0.436661    1.547532
14                     Pets     -0.222189    0.800764

In [31]:
# sort by highest odds ratio first
summary_table.sort_values(by=['Odds Ratio'], ascending=False)

Feature Names  Coefficients  Odds Ratio
3                  Reason 3      2.903267   18.233624
1                  Reason 1      2.880280   17.819257
2                  Reason 2      0.923240    2.517435
4                  Reason 4      0.889674    2.434336
7    Transportation Expense      0.599277    1.820801
13                 Children      0.436661    1.547532
11          Body Mass Index      0.247163    1.280388
12                Education      0.211180    1.235135
5               Month Value      0.070055    1.072567
8          Distance to Work      0.029727    1.030173
10  Daily Work Load Average      0.013103    1.013189
9                       Age     -0.183230    0.832576
14                     Pets     -0.222189    0.800764
6           Day of the Week     -0.295224    0.744365
0                 Intercept     -1.723676    0.178409

A feature is not particularly important if its:
- coefficient is around 0
- odds ratio is around 1

A weight of 0 implies that no matter the feature value, we will multiply it by 0 in the model.

For a unit change in the standardized feature, the odds increase by a multiple equal to the odds ratio. Hence, 1 implies no change.

I have assumed at least a 5% change in odds to be significant.

From the above table, it seems that the `Daily Work Load Average` and `Distance to Work` make no difference, given all features.

According to the table, if people have poisoning (Reason 3), they are 18.23 times more likely to be absent, followed by diseases (Reason 1) at 17.8 times, pregnancy and giving birth (Reason 2) at 2.5 times and light diseases (Reason 4) at 2.4 times.

It is understandable as if a person is poisoned or sick, he/she is more likely to miss work. In the case of pregnancy, a woman might miss work when she visits the gynaecologist, and when she gives birth (which is often paid leave and not absence).

Coming towards the bottom of the table, we can see that a 1 standard deviation increase in Age, Pets and Day of the Week causes the odds of missing work to go down by 17%, 20% and 26% respectively.

It could be interpreted that as a person becomes old (or senior), he/she has a higher responsibility at work and chooses to be absent less frequently.

In case of pets, when someone has a lot of pets, he/she might have hire someone to take care of them for which they don't have to miss work if the pets need to visit the vet.

Also, as we enter into a new week, people might take an extended weekend and be absent, but as we progress towards the middle of the week, absence is less frequent.

## Remove Unnecessary Features

Now we will remove the unnecessary features, split the dataset and rebuild the classification model.

In [32]:
scaled_inputs

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1     0.182726        -0.683704   
1           0         0         0         0     0.182726        -0.683704   
2           0         0         0         1     0.182726        -0.007725   
3           1         0         0         0     0.182726         0.668253   
4           0         0         0         1     0.182726         0.668253   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0    -0.388293        -0.007725   
696         1         0         0         0    -0.388293        -0.007725   
697         1         0         0         0    -0.388293         0.668253   
698         0         0         0         1    -0.388293         0.668253   
699         0         0         0         1    -0.388293         0.668253   

     Transportation Expense  Distance to Work       Age  \
0                  1.005844          0.412816 -0.536062   
1                 -1.574681         -1.141882  2.130803   
2                 -0.654143          1.426749  0.248310   
3                  0.854936         -1.682647  0.405184   
4                  1.005844          0.412816 -0.536062   
..                      ...               ...       ...   
695               -0.654143         -0.533522  0.562059   
696                0.040034         -0.263140 -1.320435   
697                1.624567         -0.939096 -1.320435   
698                0.190942         -0.939096 -0.692937   
699                1.036026          0.074838  0.562059   

     Daily Work Load Average  Body Mass Index  Education  Children      Pets  
0                  -0.806331         0.767431          0  0.880469  0.268487  
1                  -0.806331         1.002633          0 -0.019280 -0.589690  
2                  -0.806331         1.002633          0 -0.919030 -0.589690  
3                  -0.806331        -0.643782          0  0.880469 -0.589690  
4                  -0.806331         0.767431          0  0.880469  0.268487  
..                       ...              ...        ...       ...       ...  
695                -0.853789        -1.114186          1  0.880469 -0.589690  
696                -0.853789        -0.643782          0 -0.019280  1.126663  
697                -0.853789        -0.408580          1 -0.919030 -0.589690  
698                -0.853789        -0.408580          1 -0.919030 -0.589690  
699                -0.853789        -0.408580          0 -0.019280  0.268487  

[700 rows x 14 columns]

In [33]:
# remove insignificant features
# update the inputs
scaled_inputs_mod = scaled_inputs.drop(['Distance to Work'], axis=1)
scaled_inputs_mod = scaled_inputs_mod.drop(['Daily Work Load Average'], axis=1)

In [34]:
scaled_inputs_mod

Reason 1  Reason 2  Reason 3  Reason 4  Month Value  Day of the Week  \
0           0         0         0         1     0.182726        -0.683704   
1           0         0         0         0     0.182726        -0.683704   
2           0         0         0         1     0.182726        -0.007725   
3           1         0         0         0     0.182726         0.668253   
4           0         0         0         1     0.182726         0.668253   
..        ...       ...       ...       ...          ...              ...   
695         1         0         0         0    -0.388293        -0.007725   
696         1         0         0         0    -0.388293        -0.007725   
697         1         0         0         0    -0.388293         0.668253   
698         0         0         0         1    -0.388293         0.668253   
699         0         0         0         1    -0.388293         0.668253   

     Transportation Expense       Age  Body Mass Index  Education  Children  \
0                  1.005844 -0.536062         0.767431          0  0.880469   
1                 -1.574681  2.130803         1.002633          0 -0.019280   
2                 -0.654143  0.248310         1.002633          0 -0.919030   
3                  0.854936  0.405184        -0.643782          0  0.880469   
4                  1.005844 -0.536062         0.767431          0  0.880469   
..                      ...       ...              ...        ...       ...   
695               -0.654143  0.562059        -1.114186          1  0.880469   
696                0.040034 -1.320435        -0.643782          0 -0.019280   
697                1.624567 -1.320435        -0.408580          1 -0.919030   
698                0.190942 -0.692937        -0.408580          1 -0.919030   
699                1.036026  0.562059        -0.408580          0 -0.019280   

         Pets  
0    0.268487  
1   -0.589690  
2   -0.589690  
3   -0.589690  
4    0.268487  
..        ...  
695 -0.589690  
696  1.126663  
697 -0.589690  
698 -0.589690  
699  0.268487  

[700 rows x 12 columns]

### Split the Dataset

In [35]:
# split the data into 80:20
# use random state = 1
# stratify the data
x_train, x_test, y_train, y_test = train_test_split(scaled_inputs_mod, targets, test_size=0.2, stratify=targets, random_state=1)

### Build A New Model

In [36]:
# build a logistic regression model
new_reg = LogisticRegression()

# train the model
new_reg.fit(x_train, y_train)

LogisticRegression()

In [37]:
new_reg.score(x_train, y_train)

0.7821428571428571

### Create New Summary Table

In [38]:
# create a dataframe for features and coefficients
summary_table = pd.DataFrame(data = scaled_inputs_mod.columns.values, columns = ['Feature Names'])
# add the coeff
# transpose the coeff as it has multiple columns
summary_table['Coefficients'] = np.transpose(new_reg.coef_)
# add the intercept
# shift all rows 1 down and add at the beginning
summary_table.index = summary_table.index + 1
summary_table.loc[0] = ['Intercept', new_reg.intercept_[0]]
# calculate the odds ratio
# append in the summary table
summary_table['Odds Ratio'] = np.exp(summary_table['Coefficients'])
# sort by highest odds ratio first
summary_table.sort_values(by=['Odds Ratio'], ascending=False)

Feature Names  Coefficients  Odds Ratio
3                 Reason 3      2.905188   18.268682
1                 Reason 1      2.884277   17.890630
2                 Reason 2      0.930553    2.535912
4                 Reason 4      0.892699    2.441712
7   Transportation Expense      0.606298    1.833631
11                Children      0.434739    1.544560
9          Body Mass Index      0.249953    1.283965
10               Education      0.183126    1.200965
5              Month Value      0.067504    1.069834
8                      Age     -0.189029    0.827763
12                    Pets     -0.222479    0.800532
6          Day of the Week     -0.291153    0.747401
0                Intercept     -1.721328    0.178828